tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects.ts

```diff
@@ -62,9 +139,152 @@ export class DebuggerEffects {
+        );
+      }),
+      tap(() => this.store.dispatch(executionDigestsRequested())),
+      mergeMap(([props, runId, displayCount, _]) => {
+        const begin = 0;
+        const end = Math.min(props.numExecutions, displayCount);
+        return this.dataSource.fetchExecutionDigests(runId!, begin, end).pipe(
+          map((digests) => {
+            return executionDigestsLoaded(digests);
+          })
+        );
+        // TODO(cais): Add catchError() to pipe.
+      })
+    )
+  );
+
+  /** @export */
+  readonly loadExecutionDigests$ = createEffect(() =>
+    this.actions$.pipe(
+      ofType(requestExecutionDigests),
+      withLatestFrom(this.store.select(getExecutionDigestsLoaded)),
+      filter(([_, loaded]) => loaded.state !== DataLoadState.LOADING),
+      map(([props, loaded]) => {
+        return {
+          props,
+          loaded,
+          missingPages: getMissingPages(
+            props.begin,
+            props.end,
+            props.pageSize,
+            loaded.numExecutions,
+            loaded.pageLoadedSizes
+          ),
+        };
+      }),
+      filter(({missingPages}) => missingPages.length > 0),
+      tap(() => this.store.dispatch(executionDigestsRequested())),
+      withLatestFrom(this.store.select(getExecutionDigestsLoaded)),
+      mergeMap(([{props, missingPages}, loaded]) => {
+        const {runId, pageSize} = props;
+        const actualBegin = missingPages[0] * pageSize;
+        const actualEnd = Math.min(
+          loaded.numExecutions,
+          (missingPages[missingPages.length - 1] + 1) * pageSize
+        );
+        return this.dataSource
+          .fetchExecutionDigests(runId, actualBegin, actualEnd)
+          .pipe(
+            map(
+              (digests) => {
+                return executionDigestsLoaded(digests);
+              }
+              // TODO(cais): Add catchError() to pipe.
+            )
+          );
+      })
+    )
+  );
+
+  /** @export */
+  readonly executionScrollTriggeredLoading$ = createEffect(() =>
+    this.actions$.pipe(
+      ofType(executionScrollLeft, executionScrollRight),
+      withLatestFrom(
+        this.store.select(getActiveRunId),
+        this.store.select(getExecutionScrollBeginIndex),
+        this.store.select(getNumExecutions),
+        this.store.select(getDisplayCount),
+        this.store.select(getExecutionPageSize)
+      ),
+      filter((data) => {
+        const runId = data[1];
+        return runId !== null;
+      }),
+      map(
+        ([
+          _,
+          runId,
+          scrollBeginIndex,
+          numExecutions,
+          displayCount,
+          pageSize,
+        ]) => {
+          const begin = scrollBeginIndex;
+          const end = Math.min(numExecutions, begin + displayCount);
+          return requestExecutionDigests({
+            runId: runId!,
+            begin,
+            end,
+            pageSize,
+          });
+        }
+      )
+    )
+  );
```

Comment:
I would like to propose major change to this block of code. These two effects are interacting with each other using whole flux cycles. Instead, how about this?

```ts

   private readonly digestRequired$ = this.actions$.pipe(
    ofType(executionScrollLeft, executionScrollRight),
    withLatestFrom(
      this.store.select(getActiveRunId),
      this.store.select(getExecutionScrollBeginIndex),
      this.store.select(getNumExecutions),
      this.store.select(getDisplayCount),
      this.store.select(getExecutionPageSize)
    ),
    filter(data => {
      const runId = data[1];
      return runId !== null;
    }),
    map(
      ([_, runId, scrollBeginIndex, numExecutions, displayCount, pageSize]) => {
        const begin = scrollBeginIndex;
        const end = Math.min(numExecutions, begin + displayCount);
        return {
          runId: runId!,
          begin,
          end,
          pageSize
        };
      }
    )
  );

  readonly loadExecutionDigests$ = createEffect(() =>
    digestRequired$.pipe(
      withLatestFrom(this.store.select(getExecutionDigestsLoaded)),
      filter(([_, loaded]) => loaded.state !== DataLoadState.LOADING),
      map(([props, loaded]) => {
        return {
          props,
          loaded,
          missingPages: getMissingPages(
            props.begin,
            props.end,
            props.pageSize,
            loaded.numExecutions,
            loaded.pageLoadedSizes
          )
        };
      }),
      filter(({ missingPages }) => missingPages.length > 0),
      tap(() => this.store.dispatch(executionDigestsRequested())),
      withLatestFrom(this.store.select(getExecutionDigestsLoaded)),
      mergeMap(([{ props, missingPages }, loaded]) => {
        const { runId, pageSize } = props;
        const actualBegin = missingPages[0] * pageSize;
        const actualEnd = Math.min(
          loaded.numExecutions,
          (missingPages[missingPages.length - 1] + 1) * pageSize
        );
        return this.dataSource
          .fetchExecutionDigests(runId, actualBegin, actualEnd)
          .pipe(
            map(
              digests => {
                return executionDigestsLoaded(digests);
              }
              // TODO(cais): Add catchError() to pipe.
            )
          );
      })
    )
  );

  // ALTERNATIVE
  readonly loadExecutionDigests2$ = createEffect(() =>
    this.actions$.pipe(
      ofType(executionScrollLeft, executionScrollRight),
      withLatestFrom(
        this.store.select(getExecutionDigestsLoaded),
        this.store.select(getActiveRunId),
        this.store.select(getExecutionScrollBeginIndex),
        this.store.select(getNumExecutions),
        this.store.select(getDisplayCount),
        this.store.select(getExecutionPageSize)
      ),
      filter(
        ([_, loaded, runId]) =>
          loaded.state !== DataLoadState.LOADING && runId !== null
      ),
      map(args => {
        const [
          ,
          loaded,
          runId,
          scrollBeginIndex,
          numExecutions,
          displayCount,
          pageSize
        ] = args;
        const begin = scrollBeginIndex;
        const end = Math.min(numExecutions, begin + displayCount);
        return {
          runId,
          scrollBeginIndex,
          numExecutions,
          displayCount,
          pageSize,
          loaded,
          missingPages: getMissingPages(
            begin,
            end,
            pageSize,
            loaded.numExecutions,
            loaded.pageLoadedSizes
          )
        };
      }),
      filter(({ missingPages }) => missingPages.length > 0),
      tap(() => this.store.dispatch(executionDigestsRequested())),
      mergeMap(({loaded, missingPages, runId, pageSize}) => {
        const actualBegin = missingPages[0] * pageSize;
        const actualEnd = Math.min(
          loaded.numExecutions,
          (missingPages[missingPages.length - 1] + 1) * pageSize
        );

        return this.dataSource
          .fetchExecutionDigests(runId, actualBegin, actualEnd)
          .pipe(
            map(
              digests => {
                return executionDigestsLoaded(digests);
              }
              // TODO(cais): Add catchError() to pipe.
            )
          );
      })
    )
  );
```

Reply:
Done. Adopted your first suggestion. It simplifies the code and gets ride of an action type. Thanks.